### ISSUE TITLE
Calculate loss as part of metrics when evaluating validation/test dataset

### ISSUE THREAD

**Issue by logicmd** [faneshion/MatchZoo#369]
Hi folks,

It is well known that we can use loss in training/validation dataset to check if model is overfitting or not.
Do we know how to calculate loss value as part of metrics when evaluating validation/test dataset ?
https://github.com/faneshion/MatchZoo/blob/master/matchzoo/main.py#L178

Attached a snippet of my training log

> [09-28-2018 20:45:50]	[Train:train] Iter:11	loss=0.493236
> [09-28-2018 20:45:50]	[Eval:valid] Iter:11	accuracy=0.824670
> [09-28-2018 20:46:50]	[Eval:test] Iter:11	accuracy=0.585881

**Comment by ZizhenWang** [comment]
When we calculate loss in either training or validation dataset, we use the model predict result and the ground truth. So if you have ground truth on your test dataset, you can calculate the loss and the other metrics just treat it as a training dataset.

**Comment by logicmd** [comment]
@ZizhenWang thanks for the reply. Looks right now we are using tensorflow api to calculate the loss in training dataset https://github.com/faneshion/MatchZoo/blob/master/matchzoo/main.py#L156?

Do you have any code example of calculating loss in validation dataset? I can't find any code doing that...

**Comment by ZizhenWang** [comment]
actually the following code block from line 158 to line 178 in `main.py` is the calculation of evaluate metrics (loss or other metrics) on validation dataset, you can run a model and see loggings on the screen.

**Comment by ZizhenWang** [comment]
you can set the validation dataset path in model configuration file, see demo in https://github.com/faneshion/MatchZoo/blob/ec0fef5e849f2439b08b2cf81e3192e39725eae0/examples/toy_example/config/matchpyramid_classify.config#L35

**Comment by logicmd** [comment]
@ZizhenWang Sorry, I don't quite follow that... 

> actually the following code block from line 158 to line 178 in main.py is the calculation of evaluate metrics (loss or other metrics) on validation dataset, 

How can I call `history.history['loss'][0]` in line 178? There is no `history` right?

**Comment by ZizhenWang** [comment]
It uses different ways to calculate metrics. 
While in training,  it can calculate `loss`  by default Keras `history`. 
While in validation, we first get ground truth by `data generator` https://github.com/faneshion/MatchZoo/blob/ec0fef5e849f2439b08b2cf81e3192e39725eae0/matchzoo/main.py#L158-L163
And then get model predict result
https://github.com/faneshion/MatchZoo/blob/ec0fef5e849f2439b08b2cf81e3192e39725eae0/matchzoo/main.py#L164
At last, we calculate the metrics by
https://github.com/faneshion/MatchZoo/blob/ec0fef5e849f2439b08b2cf81e3192e39725eae0/matchzoo/main.py#L165-L176
Maybe some variables in above code lead to confusion, but i think the pipeline is clear.
